In [1]:
# !pip list

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [212]:
import os
import warnings

In [4]:
os.path.abspath("__file__")

'D:\\E2EML\\E2E-ML\\research\\__file__'

In [5]:
curr_dir = os.path.dirname(os.path.abspath("__file__"))

In [6]:
curr_dir

'D:\\E2EML\\E2E-ML\\research'

In [7]:
matches_path = os.path.join(curr_dir,"..","artifacts","data_ingestion","ipl_dataset","matches.csv")
deliveries_path = os.path.join(curr_dir,"..","artifacts","data_ingestion","ipl_dataset","deliveries.csv")

In [8]:
matches_path

'D:\\E2EML\\E2E-ML\\research\\..\\artifacts\\data_ingestion\\ipl_dataset\\matches.csv'

In [9]:
deliveries_path

'D:\\E2EML\\E2E-ML\\research\\..\\artifacts\\data_ingestion\\ipl_dataset\\deliveries.csv'

In [10]:
matches_data = os.path.normpath(matches_path)
deliveries_data = os.path.normpath(deliveries_path)

In [11]:
print(matches_data)
print(deliveries_data)

D:\E2EML\E2E-ML\artifacts\data_ingestion\ipl_dataset\matches.csv
D:\E2EML\E2E-ML\artifacts\data_ingestion\ipl_dataset\deliveries.csv


In [12]:
match = pd.read_csv(matches_data)
delivery = pd.read_csv(deliveries_data)

In [13]:
match.head()

id    Season       city        date                        team1  \
0   1  IPL-2017  Hyderabad  05-04-2017          Sunrisers Hyderabad   
1   2  IPL-2017       Pune  06-04-2017               Mumbai Indians   
2   3  IPL-2017     Rajkot  07-04-2017                Gujarat Lions   
3   4  IPL-2017     Indore  08-04-2017       Rising Pune Supergiant   
4   5  IPL-2017  Bangalore  08-04-2017  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

In [14]:
delivery.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...         0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...         0   
2     3  DA Warner    S Dhawan  TS Mills              0  ...         0   
3     4  DA Warner    S Dhawan  TS Mills              0  ...         0   
4     5  DA Warner    S Dhawan  TS Mills              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs  player_dismissed dismissal_kind fielder  
0           0               NaN            NaN     NaN  
1           0               NaN            NaN     NaN  
2           4               NaN            NaN     NaN  
3           0               NaN            NaN     NaN  
4           2               NaN            NaN     NaN  

[5 rows x 21 columns]

In [15]:
match.shape

(756, 18)

In [16]:
delivery.shape

(179078, 21)

In [17]:
len(delivery["match_id"].unique())

756

In [18]:
match.columns

Index(['id', 'Season', 'city', 'date', 'team1', 'team2', 'toss_winner',
       'toss_decision', 'result', 'dl_applied', 'winner', 'win_by_runs',
       'win_by_wickets', 'player_of_match', 'venue', 'umpire1', 'umpire2',
       'umpire3'],
      dtype='object')

In [19]:
delivery.columns

Index(['match_id', 'inning', 'batting_team', 'bowling_team', 'over', 'ball',
       'batsman', 'non_striker', 'bowler', 'is_super_over', 'wide_runs',
       'bye_runs', 'legbye_runs', 'noball_runs', 'penalty_runs',
       'batsman_runs', 'extra_runs', 'total_runs', 'player_dismissed',
       'dismissal_kind', 'fielder'],
      dtype='object')

##### All the matches deliveries are present in delivery datset

In [20]:
match.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   id               756 non-null    int64 
 1   Season           756 non-null    object
 2   city             749 non-null    object
 3   date             756 non-null    object
 4   team1            756 non-null    object
 5   team2            756 non-null    object
 6   toss_winner      756 non-null    object
 7   toss_decision    756 non-null    object
 8   result           756 non-null    object
 9   dl_applied       756 non-null    int64 
 10  winner           752 non-null    object
 11  win_by_runs      756 non-null    int64 
 12  win_by_wickets   756 non-null    int64 
 13  player_of_match  752 non-null    object
 14  venue            756 non-null    object
 15  umpire1          754 non-null    object
 16  umpire2          754 non-null    object
 17  umpire3          119 non-null    ob

In [21]:
delivery.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 179078 entries, 0 to 179077
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   match_id          179078 non-null  int64 
 1   inning            179078 non-null  int64 
 2   batting_team      179078 non-null  object
 3   bowling_team      179078 non-null  object
 4   over              179078 non-null  int64 
 5   ball              179078 non-null  int64 
 6   batsman           179078 non-null  object
 7   non_striker       179078 non-null  object
 8   bowler            179078 non-null  object
 9   is_super_over     179078 non-null  int64 
 10  wide_runs         179078 non-null  int64 
 11  bye_runs          179078 non-null  int64 
 12  legbye_runs       179078 non-null  int64 
 13  noball_runs       179078 non-null  int64 
 14  penalty_runs      179078 non-null  int64 
 15  batsman_runs      179078 non-null  int64 
 16  extra_runs        179078 non-null  int

##### Null values are present in both the datasets

In [22]:
delivery[delivery["match_id"]==1]

match_id  inning                 batting_team  \
0           1       1          Sunrisers Hyderabad   
1           1       1          Sunrisers Hyderabad   
2           1       1          Sunrisers Hyderabad   
3           1       1          Sunrisers Hyderabad   
4           1       1          Sunrisers Hyderabad   
..        ...     ...                          ...   
243         1       2  Royal Challengers Bangalore   
244         1       2  Royal Challengers Bangalore   
245         1       2  Royal Challengers Bangalore   
246         1       2  Royal Challengers Bangalore   
247         1       2  Royal Challengers Bangalore   

                    bowling_team  over  ball      batsman  non_striker  \
0    Royal Challengers Bangalore     1     1    DA Warner     S Dhawan   
1    Royal Challengers Bangalore     1     2    DA Warner     S Dhawan   
2    Royal Challengers Bangalore     1     3    DA Warner     S Dhawan   
3    Royal Challengers Bangalore     1     4    DA Warner     S Dhawan   
4    Royal Challengers Bangalore     1     5    DA Warner     S Dhawan   
..                           ...   ...   ...          ...          ...   
243          Sunrisers Hyderabad    19     6    YS Chahal  A Choudhary   
244          Sunrisers Hyderabad    20     1    YS Chahal  A Choudhary   
245          Sunrisers Hyderabad    20     2    YS Chahal  A Choudhary   
246          Sunrisers Hyderabad    20     3  A Choudhary    YS Chahal   
247          Sunrisers Hyderabad    20     4  A Choudhary    YS Chahal   

          bowler  is_super_over  ...  bye_runs  legbye_runs  noball_runs  \
0       TS Mills              0  ...         0            0            0   
1       TS Mills              0  ...         0            0            0   
2       TS Mills              0  ...         0            0            0   
3       TS Mills              0  ...         0            0            0   
4       TS Mills              0  ...         0            0            0   
..           ...            ...  ...       ...          ...          ...   
243      B Kumar              0  ...         0            0            0   
244  BCJ Cutting              0  ...         0            0            0   
245  BCJ Cutting              0  ...         0            0            0   
246  BCJ Cutting              0  ...         0            0            0   
247  BCJ Cutting              0  ...         0            0            0   

     penalty_runs  batsman_runs  extra_runs  total_runs  player_dismissed  \
0               0             0           0           0               NaN   
1               0             0           0           0               NaN   
2               0             4           0           4               NaN   
3               0             0           0           0               NaN   
4               0             0           2           2               NaN   
..            ...           ...         ...         ...               ...   
243             0             1           0           1               NaN   
244             0             0           0           0               NaN   
245             0             1           0           1               NaN   
246             0             6           0           6               NaN   
247             0             0           0           0         YS Chahal   

    dismissal_kind          fielder  
0              NaN              NaN  
1              NaN              NaN  
2              NaN              NaN  
3              NaN              NaN  
4              NaN              NaN  
..             ...              ...  
243            NaN              NaN  
244            NaN              NaN  
245            NaN              NaN  
246            NaN              NaN  
247        run out  CJ Jordan (sub)  

[248 rows x 21 columns]

##### Thus we got to know that match_id 1 is match between SRH & RCB

##### Total of 248 balls played that day

##### Now let's see how much each team has scored?

In [23]:
first_match = delivery[delivery["match_id"]==1]

In [24]:
innings1 = first_match[first_match["inning"]==1]
innings2 = first_match[first_match["inning"]==2]

In [25]:
print(innings1.shape)
print(innings2.shape)

(125, 21)
(123, 21)


In [26]:
innings1.tail()

match_id  inning         batting_team                 bowling_team  over  \
120         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore    20   
121         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore    20   
122         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore    20   
123         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore    20   
124         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore    20   

     ball      batsman  non_striker     bowler  is_super_over  ...  bye_runs  \
120     2     DJ Hooda  BCJ Cutting  SR Watson              0  ...         0   
121     3  BCJ Cutting     DJ Hooda  SR Watson              0  ...         0   
122     4  BCJ Cutting     DJ Hooda  SR Watson              0  ...         0   
123     5  BCJ Cutting     DJ Hooda  SR Watson              0  ...         0   
124     6  BCJ Cutting     DJ Hooda  SR Watson              0  ...         0   

     legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
120            0            0             0             1           0   
121            0            0             0             6           0   
122            0            0             0             2           0   
123            0            0             0             0           0   
124            0            0             0             6           0   

     total_runs  player_dismissed dismissal_kind fielder  
120           1               NaN            NaN     NaN  
121           6               NaN            NaN     NaN  
122           2               NaN            NaN     NaN  
123           0               NaN            NaN     NaN  
124           6               NaN            NaN     NaN  

[5 rows x 21 columns]

##### SRH played all 20 overs

In [27]:
innings2.tail()

match_id  inning                 batting_team         bowling_team  over  \
243         1       2  Royal Challengers Bangalore  Sunrisers Hyderabad    19   
244         1       2  Royal Challengers Bangalore  Sunrisers Hyderabad    20   
245         1       2  Royal Challengers Bangalore  Sunrisers Hyderabad    20   
246         1       2  Royal Challengers Bangalore  Sunrisers Hyderabad    20   
247         1       2  Royal Challengers Bangalore  Sunrisers Hyderabad    20   

     ball      batsman  non_striker       bowler  is_super_over  ...  \
243     6    YS Chahal  A Choudhary      B Kumar              0  ...   
244     1    YS Chahal  A Choudhary  BCJ Cutting              0  ...   
245     2    YS Chahal  A Choudhary  BCJ Cutting              0  ...   
246     3  A Choudhary    YS Chahal  BCJ Cutting              0  ...   
247     4  A Choudhary    YS Chahal  BCJ Cutting              0  ...   

     bye_runs  legbye_runs  noball_runs  penalty_runs  batsman_runs  \
243         0            0            0             0             1   
244         0            0            0             0             0   
245         0            0            0             0             1   
246         0            0            0             0             6   
247         0            0            0             0             0   

     extra_runs  total_runs  player_dismissed dismissal_kind          fielder  
243           0           1               NaN            NaN              NaN  
244           0           0               NaN            NaN              NaN  
245           0           1               NaN            NaN              NaN  
246           0           6               NaN            NaN              NaN  
247           0           0         YS Chahal        run out  CJ Jordan (sub)  

[5 rows x 21 columns]

##### RCB only played 19.4 overs

In [28]:
innings1["total_runs"].dtype

dtype('int64')

In [29]:
int(innings1["total_runs"].sum())

207

In [30]:
int(innings2["total_runs"].sum())

172

##### Thus SRH won the match

##### Btw we already knew it from the match dataset

In [31]:
match.head(1)

id    Season       city        date                team1  \
0   1  IPL-2017  Hyderabad  05-04-2017  Sunrisers Hyderabad   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   

   result  dl_applied               winner  win_by_runs  win_by_wickets  \
0  normal           0  Sunrisers Hyderabad           35               0   

  player_of_match                                      venue      umpire1  \
0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal  AY Dandekar   

    umpire2 umpire3  
0  NJ Llong     NaN

In [32]:
match.head(1)["winner"]

0    Sunrisers Hyderabad
Name: winner, dtype: object

##### Now how to get all the matches scores in each innings?

##### Use groupby

In [33]:
delivery.shape

(179078, 21)

In [34]:
pd.DataFrame(delivery.groupby(["match_id","inning"]))

0                                                  1
0         (1, 1)       match_id  inning         batting_team    ...
1         (1, 2)       match_id  inning                 batting_...
2         (2, 1)       match_id  inning    batting_team         ...
3         (2, 2)       match_id  inning            batting_team ...
4         (3, 1)       match_id  inning   batting_team          ...
...          ...                                                ...
1523  (11413, 2)          match_id  inning    batting_team      ...
1524  (11414, 1)          match_id  inning    batting_team      ...
1525  (11414, 2)          match_id  inning         batting_team ...
1526  (11415, 1)          match_id  inning    batting_team      ...
1527  (11415, 2)          match_id  inning         batting_team ...

[1528 rows x 2 columns]

In [35]:
delivery.groupby(["match_id","inning"]).sum()

batting_team  \
match_id inning                                                      
1        1       Sunrisers HyderabadSunrisers HyderabadSunriser...   
         2       Royal Challengers BangaloreRoyal Challengers B...   
2        1       Mumbai IndiansMumbai IndiansMumbai IndiansMumb...   
         2       Rising Pune SupergiantRising Pune SupergiantRi...   
3        1       Gujarat LionsGujarat LionsGujarat LionsGujarat...   
...                                                            ...   
11413    2       Delhi CapitalsDelhi CapitalsDelhi CapitalsDelh...   
11414    1       Delhi CapitalsDelhi CapitalsDelhi CapitalsDelh...   
         2       Chennai Super KingsChennai Super KingsChennai ...   
11415    1       Mumbai IndiansMumbai IndiansMumbai IndiansMumb...   
         2       Chennai Super KingsChennai Super KingsChennai ...   

                                                      bowling_team  over  \
match_id inning                                                            
1        1       Royal Challengers BangaloreRoyal Challengers B...  1312   
         2       Sunrisers HyderabadSunrisers HyderabadSunriser...  1259   
2        1       Rising Pune SupergiantRising Pune SupergiantRi...  1323   
         2       Mumbai IndiansMumbai IndiansMumbai IndiansMumb...  1270   
3        1       Kolkata Knight RidersKolkata Knight RidersKolk...  1280   
...                                                            ...   ...   
11413    2       Sunrisers HyderabadSunrisers HyderabadSunriser...  1301   
11414    1       Chennai Super KingsChennai Super KingsChennai ...  1294   
         2       Delhi CapitalsDelhi CapitalsDelhi CapitalsDelh...  1160   
11415    1       Chennai Super KingsChennai Super KingsChennai ...  1287   
         2       Mumbai IndiansMumbai IndiansMumbai IndiansMumb...  1295   

                 ball                                            batsman  \
match_id inning                                                            
1        1        456  DA WarnerDA WarnerDA WarnerDA WarnerDA WarnerS...   
         2        444  CH GayleMandeep SinghMandeep SinghMandeep Sing...   
2        1        455  PA PatelPA PatelPA PatelJC ButtlerPA PatelPA P...   
         2        435  AM RahaneMA AgarwalMA AgarwalMA AgarwalAM Raha...   
3        1        434  JJ RoyJJ RoyJJ RoyJJ RoyJJ RoyJJ RoyBB McCullu...   
...               ...                                                ...   
11413    2        448  P ShawP ShawS DhawanS DhawanS DhawanP ShawS Dh...   
11414    1        441  P ShawP ShawP ShawP ShawP ShawS DhawanP ShawS ...   
         2        413  F du PlessisF du PlessisF du PlessisSR WatsonS...   
11415    1        447  Q de KockRG SharmaQ de KockQ de KockQ de KockQ...   
         2        448  F du PlessisF du PlessisF du PlessisF du Pless...   

                                                       non_striker  \
match_id inning                                                      
1        1       S DhawanS DhawanS DhawanS DhawanS DhawanDA War...   
         2       Mandeep SinghCH GayleCH GayleCH GayleCH GayleC...   
2        1       JC ButtlerJC ButtlerJC ButtlerPA PatelJC Buttl...   
         2       MA AgarwalAM RahaneAM RahaneAM RahaneMA Agarwa...   
3        1       BB McCullumBB McCullumBB McCullumBB McCullumBB...   
...                                                            ...   
11413    2       S DhawanS DhawanP ShawP ShawP ShawS DhawanP Sh...   
11414    1       S DhawanS DhawanS DhawanS DhawanS DhawanP Shaw...   
         2       SR WatsonSR WatsonSR WatsonF du PlessisF du Pl...   
11415    1       RG SharmaQ de KockRG SharmaRG SharmaRG SharmaR...   
         2       SR WatsonSR WatsonSR WatsonSR WatsonF du Pless...   

                                                            bowler  \
match_id inning                                                      
1        1       TS MillsTS MillsTS MillsTS MillsTS MillsTS Mil...   
         2       A NehraA NehraA NehraA Nehr

In [36]:
delivery.groupby(["match_id","inning"]).sum()["total_runs"]

match_id  inning
1         1         207
          2         172
2         1         184
          2         187
3         1         183
                   ... 
11413     2         170
11414     1         155
          2         162
11415     1         152
          2         157
Name: total_runs, Length: 1528, dtype: int64

In [37]:
pd.DataFrame(delivery[delivery["inning"]==1].groupby(["match_id"]))

0                                                  1
0        (1,)       match_id  inning         batting_team    ...
1        (2,)       match_id  inning    batting_team         ...
2        (3,)       match_id  inning   batting_team          ...
3        (4,)       match_id  inning            batting_team ...
4        (5,)        match_id  inning                 batting...
..        ...                                                ...
751  (11347,)          match_id  inning           batting_tea...
752  (11412,)          match_id  inning         batting_team ...
753  (11413,)          match_id  inning         batting_team ...
754  (11414,)          match_id  inning    batting_team      ...
755  (11415,)          match_id  inning    batting_team      ...

[756 rows x 2 columns]

In [38]:
pd.DataFrame(delivery[delivery["inning"]==2].groupby(["match_id"]))

0                                                  1
0        (1,)       match_id  inning                 batting_...
1        (2,)       match_id  inning            batting_team ...
2        (3,)       match_id  inning           batting_team  ...
3        (4,)       match_id  inning     batting_team        ...
4        (5,)        match_id  inning      batting_team      ...
..        ...                                                ...
749  (11347,)          match_id  inning    batting_team      ...
750  (11412,)          match_id  inning    batting_team      ...
751  (11413,)          match_id  inning    batting_team      ...
752  (11414,)          match_id  inning         batting_team ...
753  (11415,)          match_id  inning         batting_team ...

[754 rows x 2 columns]

In [39]:
match.shape

(756, 18)

##### That means there are no second innings for 2 matches

In [40]:
match["dl_applied"].value_counts()

dl_applied
0    737
1     19
Name: count, dtype: int64

In [41]:
match["result"].value_counts()

result
normal       743
tie            9
no result      4
Name: count, dtype: int64

In [42]:
match[match["result"]=="no result"]

id    Season       city        date                        team1  \
300    301  IPL-2011      Delhi  21-05-2011             Delhi Daredevils   
545    546  IPL-2015  Bangalore  29-04-2015  Royal Challengers Bangalore   
570    571  IPL-2015  Bangalore  17-05-2015             Delhi Daredevils   
744  11340  IPL-2019  Bengaluru  30-04-2019  Royal Challengers Bangalore   

                           team2                  toss_winner toss_decision  \
300                Pune Warriors             Delhi Daredevils           bat   
545             Rajasthan Royals             Rajasthan Royals         field   
570  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
744             Rajasthan Royals             Rajasthan Royals         field   

        result  dl_applied winner  win_by_runs  win_by_wickets  \
300  no result           0    NaN            0               0   
545  no result           0    NaN            0               0   
570  no result           0    NaN            0               0   
744  no result           0    NaN            0               0   

    player_of_match                   venue          umpire1       umpire2  \
300             NaN        Feroz Shah Kotla        SS Hazare     RJ Tucker   
545             NaN   M Chinnaswamy Stadium        JD Cloete     PG Pathak   
570             NaN   M Chinnaswamy Stadium  HDPK Dharmasena  K Srinivasan   
744             NaN  M. Chinnaswamy Stadium      Nigel Llong  Ulhas Gandhe   

            umpire3  
300             NaN  
545             NaN  
570             NaN  
744  Anil Chaudhary

In [43]:
noResultMatchIds = match[match["result"]=="no result"]["id"].values
noResultMatchIds

array([  301,   546,   571, 11340])

In [44]:
noResultDeliveries = delivery[delivery["match_id"].isin(noResultMatchIds)]

In [45]:
noResultDeliveries.groupby(["match_id","inning"]).sum()["total_runs"]

match_id  inning
301       1          56
546       1         200
571       1         187
          2           2
11340     1          65
          2          42
Name: total_runs, dtype: int64

##### Thus match_id 301 & 546 don't have any second innings

In [46]:
delivery.groupby("match_id")["inning"]

In [47]:
singleInningMatches = delivery.groupby("match_id")["inning"].apply(lambda x:2 not in x.values)

In [48]:
singleInningMatches

match_id
1        False
2        False
3        False
4        False
5        False
         ...  
11347    False
11412    False
11413    False
11414    False
11415    False
Name: inning, Length: 756, dtype: bool

In [49]:
singleInnings = np.array(singleInningMatches[singleInningMatches.values==True].index)

In [50]:
singleInnings

array([301, 546])

In [51]:
match[match["id"].isin(singleInnings)]

id    Season       city        date                        team1  \
300  301  IPL-2011      Delhi  21-05-2011             Delhi Daredevils   
545  546  IPL-2015  Bangalore  29-04-2015  Royal Challengers Bangalore   

                team2       toss_winner toss_decision     result  dl_applied  \
300     Pune Warriors  Delhi Daredevils           bat  no result           0   
545  Rajasthan Royals  Rajasthan Royals         field  no result           0   

    winner  win_by_runs  win_by_wickets player_of_match  \
300    NaN            0               0             NaN   
545    NaN            0               0             NaN   

                     venue    umpire1    umpire2 umpire3  
300       Feroz Shah Kotla  SS Hazare  RJ Tucker     NaN  
545  M Chinnaswamy Stadium  JD Cloete  PG Pathak     NaN

##### Now let's try to merge match df and delivery df

As observed, match df doesn't contain any info regarding total runs, so let's add 1st inning total runs from delivery df to match df

In [52]:
delivery.groupby(["match_id","inning"]).sum()["total_runs"]

match_id  inning
1         1         207
          2         172
2         1         184
          2         187
3         1         183
                   ... 
11413     2         170
11414     1         155
          2         162
11415     1         152
          2         157
Name: total_runs, Length: 1528, dtype: int64

In [53]:
delivery.groupby(["match_id","inning"]).sum()["total_runs"].reset_index()

match_id  inning  total_runs
0            1       1         207
1            1       2         172
2            2       1         184
3            2       2         187
4            3       1         183
...        ...     ...         ...
1523     11413       2         170
1524     11414       1         155
1525     11414       2         162
1526     11415       1         152
1527     11415       2         157

[1528 rows x 3 columns]

In [54]:
total_scores = delivery.groupby(["match_id","inning"]).sum()["total_runs"].reset_index()

In [55]:
total_score_innings1 = total_scores[total_scores["inning"]==1]

In [56]:
total_score_innings1.head()

match_id  inning  total_runs
0         1       1         207
2         2       1         184
4         3       1         183
6         4       1         163
8         5       1         157

In [57]:
total_score_innings1 = total_score_innings1.drop(["inning"],axis=1)

In [58]:
total_score_innings1.head(2)

match_id  total_runs
0         1         207
2         2         184

##### Merge 2 dfs

In [145]:
match_totalRuns1 = match.merge(total_score_innings1, left_on="id", right_on="match_id")
match_totalRuns1.head()

id    Season       city        date                        team1  \
0   1  IPL-2017  Hyderabad  05-04-2017          Sunrisers Hyderabad   
1   2  IPL-2017       Pune  06-04-2017               Mumbai Indians   
2   3  IPL-2017     Rajkot  07-04-2017                Gujarat Lions   
3   4  IPL-2017     Indore  08-04-2017       Rising Pune Supergiant   
4   5  IPL-2017  Bangalore  08-04-2017  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  match_id  total_runs  
0     AY Dandekar       NJ Llong     NaN         1         207  
1  A Nand Kishore         S Ravi     NaN         2         184  
2     Nitin Menon      CK Nandan     NaN         3         183  
3    AK Chaudhary  C Shamshuddin     NaN         4         163  
4             NaN            NaN     NaN         5         157

In [146]:
match_totalRuns1.shape

(756, 20)

In [147]:
match.shape

(756, 18)

##### Match id's without 2nd innings will still have 1st innings total scores

In [148]:
match_totalRuns1[match_totalRuns1["id"].isin(singleInnings)]

id    Season       city        date                        team1  \
300  301  IPL-2011      Delhi  21-05-2011             Delhi Daredevils   
545  546  IPL-2015  Bangalore  29-04-2015  Royal Challengers Bangalore   

                team2       toss_winner toss_decision     result  dl_applied  \
300     Pune Warriors  Delhi Daredevils           bat  no result           0   
545  Rajasthan Royals  Rajasthan Royals         field  no result           0   

    winner  win_by_runs  win_by_wickets player_of_match  \
300    NaN            0               0             NaN   
545    NaN            0               0             NaN   

                     venue    umpire1    umpire2 umpire3  match_id  total_runs  
300       Feroz Shah Kotla  SS Hazare  RJ Tucker     NaN       301          56  
545  M Chinnaswamy Stadium  JD Cloete  PG Pathak     NaN       546         200

##### Let's check the team names

In [149]:
np.sort(match_totalRuns1["team1"].unique())

array(['Chennai Super Kings', 'Deccan Chargers', 'Delhi Capitals',
       'Delhi Daredevils', 'Gujarat Lions', 'Kings XI Punjab',
       'Kochi Tuskers Kerala', 'Kolkata Knight Riders', 'Mumbai Indians',
       'Pune Warriors', 'Rajasthan Royals', 'Rising Pune Supergiant',
       'Rising Pune Supergiants', 'Royal Challengers Bangalore',
       'Sunrisers Hyderabad'], dtype=object)

In [150]:
len(match_totalRuns1["team1"].unique())

15

In [151]:
np.sort(match_totalRuns1["team2"].unique())

array(['Chennai Super Kings', 'Deccan Chargers', 'Delhi Capitals',
       'Delhi Daredevils', 'Gujarat Lions', 'Kings XI Punjab',
       'Kochi Tuskers Kerala', 'Kolkata Knight Riders', 'Mumbai Indians',
       'Pune Warriors', 'Rajasthan Royals', 'Rising Pune Supergiant',
       'Rising Pune Supergiants', 'Royal Challengers Bangalore',
       'Sunrisers Hyderabad'], dtype=object)

In [152]:
len(match_totalRuns1["team2"].unique())

15

In [153]:
np.sort(match_totalRuns1["toss_winner"].unique())

array(['Chennai Super Kings', 'Deccan Chargers', 'Delhi Capitals',
       'Delhi Daredevils', 'Gujarat Lions', 'Kings XI Punjab',
       'Kochi Tuskers Kerala', 'Kolkata Knight Riders', 'Mumbai Indians',
       'Pune Warriors', 'Rajasthan Royals', 'Rising Pune Supergiant',
       'Rising Pune Supergiants', 'Royal Challengers Bangalore',
       'Sunrisers Hyderabad'], dtype=object)

In [154]:
len(match_totalRuns1["toss_winner"].unique())

15

In [155]:
match_totalRuns1["winner"].unique()

array(['Sunrisers Hyderabad', 'Rising Pune Supergiant',
       'Kolkata Knight Riders', 'Kings XI Punjab',
       'Royal Challengers Bangalore', 'Mumbai Indians',
       'Delhi Daredevils', 'Gujarat Lions', 'Chennai Super Kings',
       'Rajasthan Royals', 'Deccan Chargers', 'Pune Warriors',
       'Kochi Tuskers Kerala', nan, 'Rising Pune Supergiants',
       'Delhi Capitals'], dtype=object)

Nan values are present

In [156]:
match_totalRuns1["winner"] = match_totalRuns1["winner"].fillna("")

In [157]:
np.sort(match_totalRuns1["winner"].unique())

array(['', 'Chennai Super Kings', 'Deccan Chargers', 'Delhi Capitals',
       'Delhi Daredevils', 'Gujarat Lions', 'Kings XI Punjab',
       'Kochi Tuskers Kerala', 'Kolkata Knight Riders', 'Mumbai Indians',
       'Pune Warriors', 'Rajasthan Royals', 'Rising Pune Supergiant',
       'Rising Pune Supergiants', 'Royal Challengers Bangalore',
       'Sunrisers Hyderabad'], dtype=object)

In [158]:
len(match_totalRuns1["winner"].unique())

16

In [159]:
currentTeamNames = [
    "Chennai Super Kings",
    "Delhi Capitals",
    "Gujarat Titans",
    "Kolkata Knight Riders",
    "Mumbai Indians",
    "Punjab Kings",
    "Rajasthan Royals",
    "Royal Challengers Bangalore",
    "Sunrisers Hyderabad"
]

In [160]:
def updateTeamNames(df,oldTeamName,newTeamName):
    df["team1"] = df["team1"].str.replace(oldTeamName, newTeamName)
    df["team2"] = df["team2"].str.replace(oldTeamName, newTeamName)
    df["toss_winner"] = df["toss_winner"].str.replace(oldTeamName, newTeamName)
    df["winner"] = df["winner"].str.replace(oldTeamName, newTeamName)

In [161]:
updateTeamNames(match_totalRuns1, "Deccan Chargers", "Sunrisers Hyderabad")
updateTeamNames(match_totalRuns1, "Gujarat Lions", "Gujarat Titans")
updateTeamNames(match_totalRuns1, "Rising Pune Supergiant", "Chennai Super Kings")
updateTeamNames(match_totalRuns1, "Chennai Super Kingss", "Chennai Super Kings")
updateTeamNames(match_totalRuns1, "Delhi Daredevils", "Delhi Capitals")
updateTeamNames(match_totalRuns1, "Kings XI Punjab", "Punjab Kings")

In [162]:
match_totalRuns1.shape

(756, 20)

In [163]:
match_totalRuns1["team1"].unique()

array(['Sunrisers Hyderabad', 'Mumbai Indians', 'Gujarat Titans',
       'Chennai Super Kings', 'Royal Challengers Bangalore',
       'Kolkata Knight Riders', 'Delhi Capitals', 'Punjab Kings',
       'Rajasthan Royals', 'Kochi Tuskers Kerala', 'Pune Warriors'],
      dtype=object)

In [164]:
len(match_totalRuns1["team1"].unique())

11

In [165]:
match_totalRuns1["team1"].value_counts()

team1
Sunrisers Hyderabad            106
Chennai Super Kings            104
Mumbai Indians                 101
Punjab Kings                    91
Royal Challengers Bangalore     85
Kolkata Knight Riders           83
Delhi Capitals                  78
Rajasthan Royals                67
Pune Warriors                   20
Gujarat Titans                  14
Kochi Tuskers Kerala             7
Name: count, dtype: int64

In [166]:
match_totalRuns1["team2"].unique()

array(['Royal Challengers Bangalore', 'Chennai Super Kings',
       'Kolkata Knight Riders', 'Punjab Kings', 'Delhi Capitals',
       'Sunrisers Hyderabad', 'Mumbai Indians', 'Gujarat Titans',
       'Rajasthan Royals', 'Pune Warriors', 'Kochi Tuskers Kerala'],
      dtype=object)

In [167]:
len(match_totalRuns1["team2"].unique())

11

In [168]:
match_totalRuns1["team2"].value_counts()

team2
Delhi Capitals                 99
Kolkata Knight Riders          95
Royal Challengers Bangalore    95
Chennai Super Kings            90
Mumbai Indians                 86
Punjab Kings                   85
Rajasthan Royals               80
Sunrisers Hyderabad            77
Pune Warriors                  26
Gujarat Titans                 16
Kochi Tuskers Kerala            7
Name: count, dtype: int64

In [169]:
match_totalRuns1.shape

(756, 20)

In [170]:
match_totalRuns1["toss_winner"].unique()

array(['Royal Challengers Bangalore', 'Chennai Super Kings',
       'Kolkata Knight Riders', 'Punjab Kings', 'Sunrisers Hyderabad',
       'Mumbai Indians', 'Gujarat Titans', 'Delhi Capitals',
       'Rajasthan Royals', 'Kochi Tuskers Kerala', 'Pune Warriors'],
      dtype=object)

In [171]:
len(match_totalRuns1["toss_winner"].unique())

11

In [172]:
match_totalRuns1["toss_winner"].value_counts()

toss_winner
Chennai Super Kings            102
Mumbai Indians                  98
Kolkata Knight Riders           92
Delhi Capitals                  90
Sunrisers Hyderabad             89
Royal Challengers Bangalore     81
Punjab Kings                    81
Rajasthan Royals                80
Pune Warriors                   20
Gujarat Titans                  15
Kochi Tuskers Kerala             8
Name: count, dtype: int64

In [173]:
match_totalRuns1["winner"].unique()

array(['Sunrisers Hyderabad', 'Chennai Super Kings',
       'Kolkata Knight Riders', 'Punjab Kings',
       'Royal Challengers Bangalore', 'Mumbai Indians', 'Delhi Capitals',
       'Gujarat Titans', 'Rajasthan Royals', 'Pune Warriors',
       'Kochi Tuskers Kerala', ''], dtype=object)

In [174]:
len(match_totalRuns1["winner"].unique())

12

In [175]:
match_totalRuns1["winner"].value_counts()

winner
Chennai Super Kings            115
Mumbai Indians                 109
Kolkata Knight Riders           92
Sunrisers Hyderabad             87
Royal Challengers Bangalore     84
Punjab Kings                    82
Delhi Capitals                  77
Rajasthan Royals                75
Gujarat Titans                  13
Pune Warriors                   12
Kochi Tuskers Kerala             6
                                 4
Name: count, dtype: int64

In [176]:
match_totalRuns1["team1"].value_counts()

team1
Sunrisers Hyderabad            106
Chennai Super Kings            104
Mumbai Indians                 101
Punjab Kings                    91
Royal Challengers Bangalore     85
Kolkata Knight Riders           83
Delhi Capitals                  78
Rajasthan Royals                67
Pune Warriors                   20
Gujarat Titans                  14
Kochi Tuskers Kerala             7
Name: count, dtype: int64

In [177]:
len(currentTeamNames)

9

In [183]:
match_totalRuns1 = match_totalRuns1[match_totalRuns1["team1"].isin(currentTeamNames)]
match_totalRuns1 = match_totalRuns1[match_totalRuns1["team2"].isin(currentTeamNames)]

In [184]:
match_totalRuns1.shape

(697, 20)

In [185]:
match_totalRuns1 = match_totalRuns1[match_totalRuns1["toss_winner"].isin(currentTeamNames)]
match_totalRuns1 = match_totalRuns1[match_totalRuns1["winner"].isin(currentTeamNames)]

In [186]:
match_totalRuns1.shape

(694, 20)

In [188]:
match_totalRuns1.sample(n=5)

id    Season       city        date                        team1  \
28    29  IPL-2017       Pune  26-04-2017          Chennai Super Kings   
404  405  IPL-2013  Hyderabad  19-04-2013                 Punjab Kings   
177  178  IPL-2010    Kolkata  14-03-2010  Royal Challengers Bangalore   
541  542  IPL-2015    Chennai  25-04-2015          Chennai Super Kings   
429  430  IPL-2013     Mumbai  05-05-2013               Mumbai Indians   

                     team2            toss_winner toss_decision  result  \
28   Kolkata Knight Riders  Kolkata Knight Riders         field  normal   
404    Sunrisers Hyderabad           Punjab Kings           bat  normal   
177  Kolkata Knight Riders  Kolkata Knight Riders         field  normal   
541           Punjab Kings    Chennai Super Kings           bat  normal   
429    Chennai Super Kings         Mumbai Indians           bat  normal   

     dl_applied                 winner  win_by_runs  win_by_wickets  \
28            0  Kolkata Knight Riders            0               7   
404           0    Sunrisers Hyderabad            0               5   
177           0  Kolkata Knight Riders            0               7   
541           0    Chennai Super Kings           97               0   
429           0         Mumbai Indians           60               0   

    player_of_match                                      venue  \
28       RV Uthappa    Maharashtra Cricket Association Stadium   
404       GH Vihari  Rajiv Gandhi International Stadium, Uppal   
177       MK Tiwary                               Eden Gardens   
541     BB McCullum            MA Chidambaram Stadium, Chepauk   
429      MG Johnson                           Wankhede Stadium   

             umpire1        umpire2 umpire3  match_id  total_runs  
28       AY Dandekar       NJ Llong     NaN        29         182  
404  HDPK Dharmasena      CK Nandan     NaN       405         123  
177  HDPK Dharmasena      AM Saheba     NaN       178         135  
541        JD Cloete  C Shamshuddin     NaN       542         192  
429  HDPK Dharmasena      CK Nandan     NaN       430         139

Let's only consider matches with no DuckWorth Lewis applied

In [189]:
match_totalRuns1 = match_totalRuns1[match_totalRuns1["dl_applied"]==0]
match_totalRuns1 = match_totalRuns1[["match_id","city","winner","total_runs"]]

In [190]:
match_totalRuns1.shape

(676, 4)

In [191]:
match_totalRuns1.sample(n=5)

match_id       city                       winner  total_runs
209       210     Nagpur             Rajasthan Royals         159
678      7936     Jaipur             Rajasthan Royals         188
632       633  Bangalore  Royal Challengers Bangalore         158
651      7909     Mohali                 Punjab Kings         201
595       596       Pune        Kolkata Knight Riders         160

In [194]:
delivery.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...         0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...         0   
2     3  DA Warner    S Dhawan  TS Mills              0  ...         0   
3     4  DA Warner    S Dhawan  TS Mills              0  ...         0   
4     5  DA Warner    S Dhawan  TS Mills              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs  player_dismissed dismissal_kind fielder  
0           0               NaN            NaN     NaN  
1           0               NaN            NaN     NaN  
2           4               NaN            NaN     NaN  
3           0               NaN            NaN     NaN  
4           2               NaN            NaN     NaN  

[5 rows x 21 columns]

Let's merge, match_totalRuns1 (First innings total runs) and delivery

In [197]:
delivery.shape

(179078, 21)

In [200]:
match_totalRuns1.shape

(676, 4)

In [201]:
delivery_df_all = match_totalRuns1.merge(delivery, on="match_id")
delivery_df_all.head()

match_id       city               winner  total_runs_x  inning  \
0         1  Hyderabad  Sunrisers Hyderabad           207       1   
1         1  Hyderabad  Sunrisers Hyderabad           207       1   
2         1  Hyderabad  Sunrisers Hyderabad           207       1   
3         1  Hyderabad  Sunrisers Hyderabad           207       1   
4         1  Hyderabad  Sunrisers Hyderabad           207       1   

          batting_team                 bowling_team  over  ball    batsman  \
0  Sunrisers Hyderabad  Royal Challengers Bangalore     1     1  DA Warner   
1  Sunrisers Hyderabad  Royal Challengers Bangalore     1     2  DA Warner   
2  Sunrisers Hyderabad  Royal Challengers Bangalore     1     3  DA Warner   
3  Sunrisers Hyderabad  Royal Challengers Bangalore     1     4  DA Warner   
4  Sunrisers Hyderabad  Royal Challengers Bangalore     1     5  DA Warner   

   ... bye_runs legbye_runs  noball_runs  penalty_runs  batsman_runs  \
0  ...        0           0            0             0             0   
1  ...        0           0            0             0             0   
2  ...        0           0            0             0             4   
3  ...        0           0            0             0             0   
4  ...        0           0            0             0             0   

   extra_runs  total_runs_y  player_dismissed  dismissal_kind  fielder  
0           0             0               NaN             NaN      NaN  
1           0             0               NaN             NaN      NaN  
2           0             4               NaN             NaN      NaN  
3           0             0               NaN             NaN      NaN  
4           2             2               NaN             NaN      NaN  

[5 rows x 24 columns]

In [202]:
delivery_df_all.shape

(161907, 24)

In [203]:
delivery_innings2 = delivery_df_all[delivery_df_all["inning"]==2]
delivery_innings2.shape

(78481, 24)

In [204]:
delivery_innings2.head()

match_id       city               winner  total_runs_x  inning  \
125         1  Hyderabad  Sunrisers Hyderabad           207       2   
126         1  Hyderabad  Sunrisers Hyderabad           207       2   
127         1  Hyderabad  Sunrisers Hyderabad           207       2   
128         1  Hyderabad  Sunrisers Hyderabad           207       2   
129         1  Hyderabad  Sunrisers Hyderabad           207       2   

                    batting_team         bowling_team  over  ball  \
125  Royal Challengers Bangalore  Sunrisers Hyderabad     1     1   
126  Royal Challengers Bangalore  Sunrisers Hyderabad     1     2   
127  Royal Challengers Bangalore  Sunrisers Hyderabad     1     3   
128  Royal Challengers Bangalore  Sunrisers Hyderabad     1     4   
129  Royal Challengers Bangalore  Sunrisers Hyderabad     1     5   

           batsman  ... bye_runs legbye_runs  noball_runs  penalty_runs  \
125       CH Gayle  ...        0           0            0             0   
126  Mandeep Singh  ...        0           0            0             0   
127  Mandeep Singh  ...        0           0            0             0   
128  Mandeep Singh  ...        0           0            0             0   
129  Mandeep Singh  ...        0           0            0             0   

     batsman_runs  extra_runs  total_runs_y  player_dismissed  dismissal_kind  \
125             1           0             1               NaN             NaN   
126             0           0             0               NaN             NaN   
127             0           0             0               NaN             NaN   
128             2           0             2               NaN             NaN   
129             4           0             4               NaN             NaN   

     fielder  
125      NaN  
126      NaN  
127      NaN  
128      NaN  
129      NaN  

[5 rows x 24 columns]

Let's add current score column

In [207]:
delivery_innings2["total_runs_y"].dtype

dtype('int64')

In [210]:
pd.DataFrame(delivery_innings2.groupby("match_id")["total_runs_y"])

0                                                  1
0        1  125    1
126    0
127    0
128    2
129    4
 ...
1        2  373    1
374    4
375    1
376    1
377    0
 ...
2        3  617    1
618    0
619    1
620    0
621    1
 ...
3        4  838    0
839    0
840    4
841    1
842    1
 ...
4        5  1084    0
1085    1
1086    0
1087    1
1088  ...
..     ...                                                ...
671  11347  160831    0
160832    0
160833    6
160834    ...
672  11412  161055    4
161056    0
161057    0
161058    ...
673  11413  161296    0
161297    1
161298    4
161299    ...
674  11414  161543    0
161544    0
161545    1
161546    ...
675  11415  161783    0
161784    0
161785    4
161786    ...

[676 rows x 2 columns]

In [213]:
warnings.filterwarnings("ignore")

In [214]:
delivery_innings2["current_score"] = delivery_innings2.groupby("match_id")["total_runs_y"].cumsum()

In [215]:
delivery_innings2.head()

match_id       city               winner  total_runs_x  inning  \
125         1  Hyderabad  Sunrisers Hyderabad           207       2   
126         1  Hyderabad  Sunrisers Hyderabad           207       2   
127         1  Hyderabad  Sunrisers Hyderabad           207       2   
128         1  Hyderabad  Sunrisers Hyderabad           207       2   
129         1  Hyderabad  Sunrisers Hyderabad           207       2   

                    batting_team         bowling_team  over  ball  \
125  Royal Challengers Bangalore  Sunrisers Hyderabad     1     1   
126  Royal Challengers Bangalore  Sunrisers Hyderabad     1     2   
127  Royal Challengers Bangalore  Sunrisers Hyderabad     1     3   
128  Royal Challengers Bangalore  Sunrisers Hyderabad     1     4   
129  Royal Challengers Bangalore  Sunrisers Hyderabad     1     5   

           batsman  ... legbye_runs noball_runs  penalty_runs  batsman_runs  \
125       CH Gayle  ...           0           0             0             1   
126  Mandeep Singh  ...           0           0             0             0   
127  Mandeep Singh  ...           0           0             0             0   
128  Mandeep Singh  ...           0           0             0             2   
129  Mandeep Singh  ...           0           0             0             4   

     extra_runs  total_runs_y  player_dismissed  dismissal_kind  fielder  \
125           0             1               NaN             NaN      NaN   
126           0             0               NaN             NaN      NaN   
127           0             0               NaN             NaN      NaN   
128           0             2               NaN             NaN      NaN   
129           0             4               NaN             NaN      NaN   

     current_score  
125              1  
126              1  
127              1  
128              3  
129              7  

[5 rows x 25 columns]

Let's add runs left

In [216]:
delivery_innings2["runs_left"] = delivery_innings2["total_runs_x"]-delivery_innings2["current_score"]

In [217]:
delivery_innings2.head()

match_id       city               winner  total_runs_x  inning  \
125         1  Hyderabad  Sunrisers Hyderabad           207       2   
126         1  Hyderabad  Sunrisers Hyderabad           207       2   
127         1  Hyderabad  Sunrisers Hyderabad           207       2   
128         1  Hyderabad  Sunrisers Hyderabad           207       2   
129         1  Hyderabad  Sunrisers Hyderabad           207       2   

                    batting_team         bowling_team  over  ball  \
125  Royal Challengers Bangalore  Sunrisers Hyderabad     1     1   
126  Royal Challengers Bangalore  Sunrisers Hyderabad     1     2   
127  Royal Challengers Bangalore  Sunrisers Hyderabad     1     3   
128  Royal Challengers Bangalore  Sunrisers Hyderabad     1     4   
129  Royal Challengers Bangalore  Sunrisers Hyderabad     1     5   

           batsman  ... noball_runs penalty_runs  batsman_runs  extra_runs  \
125       CH Gayle  ...           0            0             1           0   
126  Mandeep Singh  ...           0            0             0           0   
127  Mandeep Singh  ...           0            0             0           0   
128  Mandeep Singh  ...           0            0             2           0   
129  Mandeep Singh  ...           0            0             4           0   

     total_runs_y  player_dismissed  dismissal_kind  fielder  current_score  \
125             1               NaN             NaN      NaN              1   
126             0               NaN             NaN      NaN              1   
127             0               NaN             NaN      NaN              1   
128             2               NaN             NaN      NaN              3   
129             4               NaN             NaN      NaN              7   

     runs_left  
125        206  
126        206  
127        206  
128        204  
129        200  

[5 rows x 26 columns]

Let's add balls left

In [218]:
delivery_innings2["balls_left"] = 126 - (delivery_innings2["over"]*6 + delivery_innings2["ball"])

In [219]:
delivery_innings2.head()

match_id       city               winner  total_runs_x  inning  \
125         1  Hyderabad  Sunrisers Hyderabad           207       2   
126         1  Hyderabad  Sunrisers Hyderabad           207       2   
127         1  Hyderabad  Sunrisers Hyderabad           207       2   
128         1  Hyderabad  Sunrisers Hyderabad           207       2   
129         1  Hyderabad  Sunrisers Hyderabad           207       2   

                    batting_team         bowling_team  over  ball  \
125  Royal Challengers Bangalore  Sunrisers Hyderabad     1     1   
126  Royal Challengers Bangalore  Sunrisers Hyderabad     1     2   
127  Royal Challengers Bangalore  Sunrisers Hyderabad     1     3   
128  Royal Challengers Bangalore  Sunrisers Hyderabad     1     4   
129  Royal Challengers Bangalore  Sunrisers Hyderabad     1     5   

           batsman  ... penalty_runs batsman_runs  extra_runs  total_runs_y  \
125       CH Gayle  ...            0            1           0             1   
126  Mandeep Singh  ...            0            0           0             0   
127  Mandeep Singh  ...            0            0           0             0   
128  Mandeep Singh  ...            0            2           0             2   
129  Mandeep Singh  ...            0            4           0             4   

     player_dismissed  dismissal_kind  fielder  current_score  runs_left  \
125               NaN             NaN      NaN              1        206   
126               NaN             NaN      NaN              1        206   
127               NaN             NaN      NaN              1        206   
128               NaN             NaN      NaN              3        204   
129               NaN             NaN      NaN              7        200   

     balls_left  
125         119  
126         118  
127         117  
128         116  
129         115  

[5 rows x 27 columns]